[ML] Transforms: Move ML "Data Frame Transforms" to Kibana management section "Transforms".

walterra opened the pull request:
## Summary

Part of #45352.

Moves "Data frame transforms" from the ML plugin to its own "transform" plugin within the Kibana management section.

![image](https://user-images.githubusercontent.com/230104/66132440-5a862600-e5f5-11e9-8410-5c6b67f6985a.png)

Review hints:

- The new plugins lives in `x-pack/legacy/plugins/transform`.
- The overall wrapping code of the plugin is based on the snapshot/restore-plugin. This includes the plugin setup, shims, react-router, cluster priviliges checks, breadcrumbs.
-  snapshot/restores cluster privileges checks have been extended with the capabilities approach (e.g. `canCreateTransform`) used in ML. The cluster privileges checks are used to check general access to sections (like `TransformManagement` or `CreateTransform`). The capabilities are used to enable/disable actions within sections (e.g. `Create transform` button on transform list page).
- References to `data frame/data_frame/dataFrame` etc. have been removed or renamed to `transform*`.
- Data frame transform related menu items have been removed from the ML plugin.
- Some code pieces like required utilities or previously shared components have been copied from ML code (e.g. `StatsBar`, `JobIcon`).
- `pages` from the ML plugin have been renamed to `sections`.
- `UI Metric constants` are defined for client based telemetry based on the snapshot/restore plugin, but the actual tracking will be implemented in a follow up.
- `DocumentationLinksService` and `DocTitleService` have been brought over from snapshot/restore but are not in use yet.
- There's now `httpService` brought over from snapshot/restore which is used for license and privileges checks based on snapshot/restore code. However, there's also `http_service.ts` brought over from ML to handle the transform API. This should be unified/cleaned-up in a follow up.
- `x-pack/legacy/plugins/transform/public/shared_imports.ts` is used to reference cross-plugin imports (some brought over as is from snapshot/restore, some from ML plugin)

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3335/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3412/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3687/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/540/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/780/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/814/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/845/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/870/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/899/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/938/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1472/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1783/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1807/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1999/)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2217/)
* Commit: 89cdaefddc099e3fbfd1797829b354513014a2e9

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2242/)
* Commit: 7429b4ce35ad09b299e62bd71539ec5749aa9700

joshdover commented on x-pack/legacy/plugins/transform/server/routes/api/app.ts:
```diff
@@ -0,0 +1,99 @@
+let xpackMainPlugin: any;
+
+export function registerAppRoutes(router: Router, plugins: Plugins) {
+  xpackMainPlugin = plugins.xpack_main;
```
You should be able to pull licensing information from the New Platform licensing plugin rather than the legacy plugin. It's available via `server.newPlatform.setup.plugins.licensing`

walterra replied:
Thanks, I looked into that a bit now but will follow up in another PR. I added a note in #45352.

joshdover commented on x-pack/legacy/plugins/transform/public/shim.ts:
```diff
@@ -0,0 +1,149 @@
+  let reactRouter: HashRouter | undefined;
+
+  return {
+    core: {
```
Many of these APIs are already available in the New Platform so you should be able to use them directly by importing from `ui/new_platform` rather than their legacy counterparts.

- `http` -> `npStart.core.http`
- `chrome` -> `npStart.core.chrome` (some APIs have been moved around, so you may still need to use the legacy service for some things (such as `chrome.getInjected`).
- `notification` -> `npStart.core.notifications` (does not support fatalError at this time).
- `documentation` -> `npStart.core.docLinks`

joshdover replied:
You may want to make your shim an extension of `npStart.core`:

```js
return {
  core: {
    ..npStart.core,
     docTitle: { ... },
  }
}
```

walterra replied:
I addressed most of your suggestions in 8e0c5bd. Migrating to `npStart.core.http` I will do in a follow-up, I added a note to #45352.

joshdover commented on x-pack/legacy/plugins/transform/public/shim.ts:
```diff
@@ -0,0 +1,149 @@
+        },
+        getRouter: (): HashRouter | undefined => {
+          return reactRouter;
+        },
```
FYI routing will not be a Core concern so there will never be an API in core like this. Instead you can use the `core.application.register` API to mount your application and setup your own router.

walterra replied:
I've removed the code related to React's router here since it's done via direct imports in the rest of the code anyway. I'm leaving the `registerAngularRoute()` in as is for now, I'm not sure yet how to use `core.application.register` so I'll leave that for another time. Added a note in #45352.

joshdover commented on x-pack/legacy/plugins/transform/public/shim.ts:
```diff
@@ -0,0 +1,149 @@
+      docTitle: {
+        change: docTitle.change,
+      },
+      savedSearches: {
```
savedSearches will come via plugins rather than via core.

walterra replied:
Moved to `plugins` in e1320e6.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2292/)
* Commit: 367c9018f9d260bae942aa7f3b67563a65a1c5db

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2313/)
* Commit: fca5504db051ab4b6dbd18821cd844c981aa36da

peteharverson commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/components/transform_list/action_start.tsx:
```diff
@@ -45,55 +41,52 @@ export const StartAction: FC<StartActionProps> = ({ items, forceDisable }) => {
+      'xpack.transform.transformList.startedTransformBulkToolTip',
       {
-        defaultMessage: 'One or more selected data frame transforms is already started.',
+        defaultMessage: 'One or more selected transforms is already started.',
```
This should be `One or more selected transforms are already started.` Or could `selected` be dropped, which would make it more readable IMO - `One or more transforms are already started.`

walterra replied:
Fixed in ecc2c59.

peteharverson commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/components/transform_list/action_start.tsx:
```diff
@@ -45,55 +41,52 @@ export const StartAction: FC<StartActionProps> = ({ items, forceDisable }) => {
       {
         defaultMessage:
-          'One or more selected data frame transforms is a completed batch transform and cannot be restarted.',
+          'One or more selected transforms is a completed batch transform and cannot be restarted.',
```
Rephrase to `One or more selected transforms are completed batch transforms and cannot be restarted.` Or even drop the word `selected`?

walterra replied:
Fixed in ecc2c59.

peteharverson commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/components/transform_list/action_stop.tsx:
```diff
@@ -4,45 +4,45 @@
+      'xpack.transform.transformList.stoppedTransformBulkToolTip',
       {
-        defaultMessage: 'One or more selected data frame transforms is already stopped.',
+        defaultMessage: 'One or more selected transforms is already stopped.',
```
As before, switch `is` to `are` and consider removing `selected`

walterra replied:
Fixed in ecc2c59.

peteharverson commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/transform_management_section.tsx:
```diff
@@ -56,29 +65,42 @@ export const Page: FC = () => {
+  const onCloseModal = () => setIsSearchSelectionVisible(false);
+  const onOpenModal = () => setIsSearchSelectionVisible(true);
+
+  const fakeVisType = {
```
Could do with a comment as to what the purpose of this is `fakeVisType` is for.

walterra replied:
In 89d3398 I got rid of `fakeVisType` and implemented it the way Dima did it for the reactified page in the ML plugin: I copied the `SearchSelection` component from visualize code and adapted it so we don't have to hack our way into visualize's code.

peteharverson commented on x-pack/legacy/plugins/transform/public/app/services/http_service.ts:
```diff
@@ -0,0 +1,52 @@
+
+// service for interacting with the server
+
+import chrome from 'ui/chrome';
```
Should you be using `npStart.core.chrome` here, or pulling this from the shim?

walterra replied:
Leaving as is for now, I tried migrating it but `chrome` via new platform doesn't have `chrome.getXsrfToken()` used in this file.

peteharverson commented on x-pack/legacy/plugins/transform/public/app/services/http_service.ts:
```diff
@@ -0,0 +1,52 @@
+import chrome from 'ui/chrome';
+
+// @ts-ignore
+import { addSystemApiHeader } from 'ui/system_api';
```
Should this come via the shim?

walterra replied:
I'll look into this in a follow up. #45352 has already a note on consolidating the somewhat mixed use of multiple `http` approaches in the whole plugin now.

peteharverson commented:
The JSON editor in the expanded row of the list is very narrow. Can it be made full width:

![image](https://user-images.githubusercontent.com/7405507/66220919-d5b70d00-e6c5-11e9-839d-78565fcd3b60.png)

peteharverson commented:
As discussed, would be good if we could get full page width content for the wizard mode. With the side nav, there is not a lot of space left for the source and preview table:

![image](https://user-images.githubusercontent.com/7405507/66221873-cd5fd180-e6c7-11e9-8533-341cd0fd0906.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2789/)
* Commit: 8e0c5bd292e2c3331990b0563ffb8c2ef441e74d

walterra commented:
@joshdover Thanks for your review! I addressed your comments and linked to commits as well as referenced the meta issue (#45352) for parts I'll address in follow-ups.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2819/)
* Commit: 4f05ebe711b33ad3ee357a00f6590d7b4fc1709d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2829/)
* Commit: e1320e673a4c59922f8cc5a590e4fea3a05585c2

pheyos commented:
x-pack ciGroup3 failures are because of tests are not robust enough against slow loading. Anomaly detection test suite is currently disabled, see #47312. Fix PR is already raised #47478.
Also, with #47323 a suite for saved search based jobs was added which still uses `cleanDataframeIndices`.

jgowdyelastic commented on x-pack/legacy/plugins/transform/public/plugin.ts:
```diff
@@ -0,0 +1,118 @@
+        $scope: any,
+        $route: any,
+        $http: ng.IHttpService,
+        $q: any,
```
`$q` is unused

walterra replied:
Fixed in 370a9f2.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/create_transform/components/wizard/wizard.tsx:
```diff
@@ -83,15 +81,22 @@ export const Wizard: SFC = React.memo(() => {
+  const [stepCreateState, setStepCreateState] = useState(getDefaultStepCreateState);
+
+  if (!isKibanaContextInitialized(kibanaContext)) {
+    // TODO proper loading indicator
```
I added a note to the list of follow-ups here #45352

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/create_transform/components/step_details/step_details_form.tsx:
```diff
@@ -75,6 +71,57 @@ export const StepDetailsForm: SFC<Props> = React.memo(({ overrides = {}, onChang
+        } catch (e) {
+          toastNotifications.addDanger(
+            i18n.translate('xpack.transform.stepDetailsForm.errorGettingTransformList', {
+              defaultMessage: 'An error occurred getting the existing transform Ids: {error}',
```
Fixed in ecc2c59.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/create_transform/components/step_details/step_details_form.tsx:
```diff
@@ -175,22 +178,22 @@ export const StepDetailsForm: SFC<Props> = React.memo(({ overrides = {}, onChang
     <EuiForm>
       <EuiFormRow
-        label={i18n.translate('xpack.ml.dataframe.stepDetailsForm.transformIdLabel', {
+        label={i18n.translate('xpack.transform.stepDetailsForm.transformIdLabel', {
```
Fixed in ecc2c59.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3051/)
* Commit: ecc2c59d841a8b14c097f5993d71bdbb932a6eb4

walterra commented:
@peteharverson I fixed the width of the JSON tab in expanded rows for both the transform plugin and the data frame analytics jobs list in be3a66c. I'll look into updating the layout in a follow up together with the feedback from Elasticsearch UI team.

walterra commented on x-pack/legacy/plugins/transform/public/app/common/transform.ts:
```diff
@@ -9,17 +9,20 @@ import { BehaviorSubject } from 'rxjs';
+export function isTransformIdValid(transformId: TransformId) {
+  return transformId.match(/^[a-z0-9\-\_]+$/g) && !transformId.match(/^([_-].*)?(.*[_-])?$/g)
+    ? true
+    : false;
```
Changed in 0e12550, also changed it for the `ml` code `isJobIdValid()`.

walterra commented on x-pack/legacy/plugins/transform/public/app/lib/authorization/components/common.ts:
```diff
@@ -0,0 +1,80 @@
+  [key: string]: string[] | undefined;
+}
+export const toArray = (value: string | string[]): string[] =>
+  Array.isArray(value) ? (value as string[]) : ([value] as string[]);
```
Fixed in b2e5e11.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/components/stats_bar/_stat.scss:
```diff
@@ -0,0 +1,8 @@
+
+}
+
+.transformStat__value {
```
Switched to `EuiHighlight` for the component in both transform/ml code in 7f5320f.

walterra replied:
Switch to just `strong` in 6b40948, totally misread the EUI docs 🤣

walterra commented on x-pack/legacy/plugins/transform/public/app/services/http_service.ts:
```diff
@@ -0,0 +1,52 @@
+        payload.body = body;
+      }
+
+      fetch(url, payload)
```
I copied this over as is from ML code, it will be revisited in a follow up because there is now a mix of approaches in this plugin with code based both on ML and snapshot/restore.

walterra commented on x-pack/legacy/plugins/transform/public/plugin.ts:
```diff
@@ -0,0 +1,114 @@
+
+const REACT_ROOT_ID = 'transformReactRoot';
+
+const template = `<kbn-management-app section="elasticsearch/transform"><div id="transformReactRoot"/></kbn-management-app>`;
```
Fixed in 370a9f2.

walterra commented on x-pack/legacy/plugins/transform/public/app/sections/transform_management/components/create_transform_button/create_transform_button.tsx:
```diff
@@ -0,0 +1,57 @@
+    !capabilities.canStartStopTransform;
+
+  const createTransformButton = (
+    <>
```
Fixed in 1b34e00.

walterra commented on x-pack/legacy/plugins/transform/public/app/services/documentation/documentation_links.ts:
```diff
@@ -0,0 +1,36 @@
+    return `${this.esPluginDocBasePath}${TRANSFORM_DOC_PATHS.plugins}`;
+  }
+
+  public getSnapshotDocUrl() {
```
You're right of course! Fixed in 1b34e00.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3287/)
* Commit: be3a66cde6289678a9d182ff249a28abcdedf9bc

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3302/)
* Commit: 6b40948d77b3287a12b2d2f3da050578dc658735

walterra commented:
@alisonelizabeth I checked with the platform team and pushed a fix for the `MaxListenersExceededWarning` you were seeing in 2f6848a.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3397/)
* Commit: 2f6848a8f851559d2e0da4cc81e238a14c97bfc7

walterra commented:
@alisonelizabeth I was able to fix the React hooks related error you were seeing in `89d3398` - the root cause was that I passed on a hook's set function as a callback to another component which was called with two arguments, thanks for pointing it out!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3422/)
* Commit: 89d3398401981b498d5a4d4f5560215e77d67c54
